### PR TITLE
SQL functions for users app

### PR DESCRIPTION
**Pull request by benrudolph**
@snopoke gave this a stab today. i think i'm doing something wrong in the test cause the SQL model is not showing up in the database. was banging my head against the computer for a while and remembered you dealt with a similar issue. was wondering if i was missing some special thing

cc: @czue @esoergel 

:tropical_fish:

### CODE REVIEW

File: corehq/apps/users/models.py [hunk 1]
```diff
@@ -19,8 +19,10 @@
 from corehq.apps.app_manager.const import USERCASE_TYPE
 from corehq.apps.domain.dbaccessors import get_docs_in_domain_by_class
 from corehq.apps.hqcase.dbaccessors import get_case_ids_in_domain_by_owner
+from couchforms.dbaccessors import get_deleted_form_ids_for_user, get_form_ids_for_user
```
Comment (dimagimon) [inline corehq/apps/users/models.py]: F401 'get_deleted_form_ids_for_user' imported but unused

Comment (dimagimon) [inline corehq/apps/users/models.py]: F401 'get_form_ids_for_user' imported but unused

File: corehq/ex-submodules/couchforms/tests/test_dbaccessors.py [hunk 2]
```diff
@@ -1,15 +1,21 @@
+import os
```
Comment (dimagimon) [inline corehq/ex-submodules/couchforms/tests/test_dbaccessors.py]: F401 'os' imported but unused

File: corehq/ex-submodules/couchforms/tests/test_dbaccessors.py [hunk 3]
```diff
@@ -1,15 +1,21 @@
...
 import datetime
 from django.test import TestCase
 
 from corehq.apps.hqadmin.dbaccessors import get_number_of_forms_in_all_domains
 from corehq.form_processor.tests.utils import FormProcessorTestUtils
-from couchforms.dbaccessors import get_forms_by_type, \
-    get_form_ids_by_type
+from corehq.util.test_utils import TestFileMixin
```
Comment (dimagimon) [inline corehq/ex-submodules/couchforms/tests/test_dbaccessors.py]: F401 'TestFileMixin' imported but unused

File: corehq/apps/users/models.py [hunk 4]
```diff
@@ -1567,24 +1568,16 @@ def get_owner_ids():
...
-        doc_ids = [r['id'] for r in db.view(view_name,
-            startkey=startkey,
-            endkey=startkey + [{}],
-            reduce=False,
-            include_docs=False,
-        )]
-        if wrap or include_docs:
-            for doc in iter_docs(db, doc_ids):
-                yield XFormInstance.wrap(doc) if wrap else doc
+        if wrap:
```
Comment (snopoke) [inline corehq/apps/users/models.py]: What happens here for the SQL case?

**Comment by snopoke** [comment]
Looks good. I'm not sure why the SQL form wouldn't show up in the DB during the test unless you're trying to see the row externally (i.e. outside the test's transaction) in which case you can't because of transaction isolation. And since the test transaction gets rolled back you'll never see the form in the DB.

File: corehq/apps/users/models.py (hunk 4 above)
Comment (benrudolph) [inline corehq/apps/users/models.py]: :+1:, still need to add that logic. got frustrated with the test not showing the xform instance haha

**Comment by benrudolph** [comment]
ok cool, wanted to make sure it wasn't that transaction issue that i was banging my head against. will dig further

**Comment by benrudolph** [comment]
Ok finally got this working. there's a fair amount of churn so may be best to review :office:

**Comment by benrudolph** [comment]
this PR does not give me hope that it'll be easy to convert the rest of HQ to SQL :disappointed:

File: submodules/ctable-src [hunk 5]
```diff
@@ -1 +1 @@
-Subproject commit f4472c121b64f722d2d259c1a533ab9010bd039d
```
Comment (snopoke) [inline submodules/ctable-src]: did you mean to commit these?

**Comment by snopoke** [comment]
Looks good. My only comment (and please don't :gun: me for only realizing this now ) is that `user.get_forms` seems to only be used when retiring/unretiring a user. If that is the case then there is probably a better way of doing what its doing for SQL other than loading all the forms and re-saving them.

I also don't think its ever called with `wrap=True`.

**Comment by benrudolph** [comment]
haha no worries, it was a good exercise. it did occur to me that it was only used in retire/unretire, but it wasn't clear to me that this wouldn't be useful. i thought it was fairly common for users to be deleted in order to remove all their forms (i.e. for test data)

also this is called with `wrap=True` it's just implicit since the default is `True` https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/users/models.py#L1664. kind of a misnomer

File: submodules/ctable-src (hunk 5 above)
Comment (benrudolph) [inline submodules/ctable-src]: nope man bad week for committing irrelevant submodules

**Comment by benrudolph** [comment]
ok talked to cory. i see now that we may be able to just drop the whole `get_forms` thing and just update the state

**Comment by snopoke** [comment]
We would still have to rebuild the cases so probably need a way to get the
form IDs which we can use to query the case transactions table.

**Comment by benrudolph** [comment]
Ok, going to take a look at the bigger picture for this. might not get to
it immediately since it's not the highest on my priority list

**Comment by snopoke** [comment]
Also fine to merge this now if you want so that it doesn't get left behind.

On 27 January 2016 at 21:09, Ben <notifications@github.com> wrote:

> Ok, going to take a look at the bigger picture for this. might not get to
> it immediately since it's not the highest on my priority list
>
> On Wed, Jan 27, 2016 at 12:32 PM, Simon Kelly <notifications@github.com>
> wrote:
>
> > We would still have to rebuild the cases so probably need a way to get
> the
> > form IDs which we can use to query the case transactions table.
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <https://github.com/dimagi/commcare-hq/pull/10001#issuecomment-175759025
> >.
>
> >
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/dimagi/commcare-hq/pull/10001#issuecomment-175800813>.
>



-- 
Simon Kelly
Senior Engineer | Dimagi South Africa

**Comment by benrudolph** [comment]
Ok @snopoke thought about this some more and think it's worth merging. i can revisist later, but there are some good cleanups, tests, and useful sql functions in this PR. i think we'll probably need a way to get forms by user at some point as well as the ability to update a form's state. in the future i'll probably just refactor the way retire/unretire works. fine to wait for tests

**Comment by benrudolph** [comment]
:white_check_mark:

**Comment by czue** [comment]
legit. :+1:

**Pull request merged**